src/threading.ts

```diff
@@ -245,15 +269,36 @@ export class BytecodeStackFrame implements IStackFrame {
     // from the previous time this method was run, and is meaningless.
     this.returnToThreadLoop = false;
 
-    // Run until we get the signal to return to the thread loop.
-    while (!this.returnToThreadLoop) {
-      var op = code.readUInt8(this.pc);
-      if (!RELEASE && logging.log_level === logging.VTRACE) {
-        vtrace(`  ${this.pc} ${annotateOpcode(op, this, code, this.pc)}`);
+    if (thread.getJVM().isJITDisabled()) {
+      // Interpret until we get the signal to return to the thread loop.
+      while (!this.returnToThreadLoop) {
+        var opCode = code.readUInt8(this.pc);
+        if (!RELEASE && logging.log_level === logging.VTRACE) {
+          vtrace(`  ${this.pc} ${annotateOpcode(op, method, code, this.pc)}`);
+        }
+        opcodeTable[opCode](thread, this, code);
+        if (!RELEASE && !this.returnToThreadLoop && logging.log_level === logging.VTRACE) {
+          vtrace(`    S: [${logging.debug_vars(this.opStack.getRaw())}], L: [${logging.debug_vars(this.locals)}]`);
+        }
       }
-      opcodeTable[op](thread, this, code);
-      if (!RELEASE && !this.returnToThreadLoop && logging.log_level === logging.VTRACE) {
-        vtrace(`    S: [${logging.debug_vars(this.opStack.getRaw())}], L: [${logging.debug_vars(this.locals)}]`);
+    } else {
+      // Run until we get the signal to return to the thread loop.
+      while (!this.returnToThreadLoop) {
+        var op = method.getOp(this.pc, code);
```

Comment:
> Any reason why your JIT still uses this getOp scheme with an array of functions instead of making a single function with a switch?

I was going to try coding it yesterday but ran out of hours :) Benchmarking and tuning the basic-block-counting changes took most of my time.

I will give it a shot today.